### PR TITLE
[PAG-1976] Sync generator tool with newest api changes

### PR DESCRIPTION
**Pull request by takumagott**
## Checklist

-   Contains unit tests ❌
-   Contains breaking changes ❌
-   Compatible with: MX 9️⃣
-   Did you update version and changelog? ❌
-   PR title properly formatted (`[XX-000]: description`)? ✅

## This PR contains

-   [ ] Bug fix
-   [ ] Feature
-   [ ] Refactor
-   [ ] Documentation
-   [X] Other (API change)

## What is the purpose of this PR?

syncs with new changes that take affect in 9.18.3+ and above 9.20.0:
- add version parameter to structure mode preview
- make datasource header for dropzones optional

corresponding docs changes: https://github.com/mendix/docs/pull/5316

## Relevant changes

see above

### CODE REVIEW

File: packages/generator-widget/generators/app/templates/pluggable/web/fullTemplateJs/src/WidgetName.editorConfig.js.ejs [hunk 1]
```diff
@@ -79,9 +79,10 @@ export function getProperties(values, defaultProperties, target) {
 // /**
 //  * @param {object} values 
 //  * @param {boolean} isDarkMode 
+//  * @param {number[]} version
 //  * @returns {object}
 //  */
-// export function getPreview(values, isDarkMode) {
+// export function getPreview(values, isDarkMode, number) {
```
Comment (iobuhov) [inline packages/generator-widget/generators/app/templates/pluggable/web/fullTemplateJs/src/WidgetName.editorConfig.js.ejs:85]:
```suggestion
// export function getPreview(values, isDarkMode, version) {
```

Comment (takumagott) [reply packages/generator-widget/generators/app/templates/pluggable/web/fullTemplateJs/src/WidgetName.editorConfig.js.ejs]: good catch. Also fixed some typings

Comment (iobuhov) [reply packages/generator-widget/generators/app/templates/pluggable/web/fullTemplateJs/src/WidgetName.editorConfig.js.ejs]: 👍

File: packages/generator-widget/generators/app/templates/pluggable/native/emptyTemplateTs/src/WidgetName.editorConfig.ts.ejs [hunk 2]
```diff
@@ -74,6 +74,8 @@ type TextProps = BaseProps & {
 type DropZoneProps = BaseProps & {
   type: "DropZone";
   property: object; // widgets property object from Values API
+  placeholder: string; // text to be shown inside the dropzone when empty
+  showDataSourceHeader: bool?; // true by default. Toggles whether to show a header containing information about the datasource
```
Comment (jeffersonRibeiro) [inline packages/generator-widget/generators/app/templates/pluggable/native/emptyTemplateTs/src/WidgetName.editorConfig.ts.ejs:78]:
```suggestion
  showDataSourceHeader?: boolean; // true by default. Toggles whether to show a header containing information about the datasource
```

**Review by jeffersonRibeiro — APPROVED** [review]

File: packages/generator-widget/generators/app/templates/pluggable/native/emptyTemplateTs/src/WidgetName.editorConfig.ts.ejs [hunk 3]
```diff
@@ -116,7 +118,7 @@ export function getProperties(_values: <%- name %>PreviewProps, defaultPropertie
 //     return errors;
 // }
 
-// export function getPreview(values: <%- name %>PreviewProps, isDarkMode: boolean): PreviewProps {
+// export function getPreview(values: <%- name %>PreviewProps, isDarkMode: boolean, version): PreviewProps {
```
Comment (tamelbrun) [inline packages/generator-widget/generators/app/templates/pluggable/native/emptyTemplateTs/src/WidgetName.editorConfig.ts.ejs:121]: The type `: number[]` is missing here.

**Review by tamelbrun — CHANGES_REQUESTED** [review]

**Comment by tamelbrun** [comment]
I'm not sure how we're going to deal with the targeted Mendix version here. These API changes were apparently important enough to backport to Mendix 9.18.3+, but we cannot release a `9.18.1` version of the `generator-widget` with features that are not supported in 9.19 without breaking our versioning scheme.

File: packages/generator-widget/generators/app/templates/pluggable/native/emptyTemplateTs/src/WidgetName.editorConfig.ts.ejs (hunk 3 above)
Comment (takumagott) [reply packages/generator-widget/generators/app/templates/pluggable/native/emptyTemplateTs/src/WidgetName.editorConfig.ts.ejs]: I hate repetitive tasks

**Comment by takumagott** [comment]
> I'm not sure how we're going to deal with the targeted Mendix version here. These API changes were apparently important enough to backport to Mendix 9.18.3+, but we cannot release a `9.18.1` version of the `generator-widget` with features that are not supported in 9.19 without breaking our versioning scheme.

Good point. As discussed in https://mendix.atlassian.net/browse/WC-1301, it's fine to publish this with 9.20 to avoid the issue

File: packages/generator-widget/generators/app/templates/pluggable/web/fullTemplateJs/src/WidgetName.editorConfig.js.ejs (hunk 1 above)
Comment (takumagott) [reply packages/generator-widget/generators/app/templates/pluggable/web/fullTemplateJs/src/WidgetName.editorConfig.js.ejs]: Can we resolve this?

Comment (tamelbrun) [reply packages/generator-widget/generators/app/templates/pluggable/web/fullTemplateJs/src/WidgetName.editorConfig.js.ejs]: Yes, we can! And I will ;)

File: packages/generator-widget/generators/app/templates/pluggable/native/emptyTemplateTs/src/WidgetName.editorConfig.ts.ejs (hunk 3 above)
Comment (tamelbrun) [reply packages/generator-widget/generators/app/templates/pluggable/native/emptyTemplateTs/src/WidgetName.editorConfig.ts.ejs]: ✨

**Comment by tamelbrun** [comment]
@takumagott Now all we're missing is an entry in the changelog :).

**Review by tamelbrun — CHANGES_REQUESTED** [review]

File: packages/generator-widget/CHANGELOG.md [hunk 4]
```diff
@@ -6,6 +6,15 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 
 ## [Unreleased]
 
+## [9.20.0] - 2022-12-01
+
+### Changed
+
+-   We now pass the Studio Pro version to the structure mode preview function
+
+-   We added the ability to make dropzone headers in structure mode previews optional
```
[the remark this answers is not in the document]
Comment (tamelbrun) [reply packages/generator-widget/CHANGELOG.md:15]: @takumagott To speed things along, here's my suggestion for an information-packed changelog:
```suggestion
- We have updated the signature of the structure mode preview function in `generator-widget` to include the Studio Pro version, which is passed in Mendix 9.20 and above (and in 9.18.3+, but not 9.19).
- We have added support to the `DropZoneProps` type in `generator-widget` for the ability to make dropzone headers in structure mode previews optional, which was added in Mendix 9.20.
```
We would like to release the 9.20 versions of both packages very soon.

File: packages/generator-widget/CHANGELOG.md [hunk 5]
```diff
@@ -6,6 +6,15 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 
 ## [Unreleased]
 
+## [9.20.0] - 2022-12-01
```
Comment (tamelbrun) [inline packages/generator-widget/CHANGELOG.md:9]: I just noticed this: the version headers in the changelog are added as part of the release process, when everything under `[Unreleased]` gets moved to a header for that specific release. Adding this header manually (with a somewhat arbitrary prediction for a release date ;)) breaks that process, please remove it :).

@jeffersonRibeiro This reminds me that I'm not seeing any release notes for `assignableTo` under `Unreleased`, we definitely need to fix that before the 9.20 release of the tools as well 🙈.

**Review by tamelbrun — CHANGES_REQUESTED** [review]

File: packages/generator-widget/CHANGELOG.md (hunk 5 above)
Comment (takumagott) [reply packages/generator-widget/CHANGELOG.md]: @tamelbrun fixed

**Review by alihcsumer — APPROVED** [review]

**Pull request merged**